s3s3/api.py

```diff
@@ -0,0 +1,114 @@
+"""
+The API for s3s3.
+"""
+import logging
+import tempfile
+
+from boto.s3.connection import S3Connection
+from boto.s3.key import Key
+
+from . import config
+
+
+if config.pubsub['redis']:
+    import redis
+    r = redis.Redis()
+else:
+    Exception('Only redis is supported right now.')
+
+
+def create_connection(connection_args):
+    """
+    Create a boto s3 connection using the connection_args dictionary.
+    ``connection_args`` A dictionary of connection arguments.
+    """
+    connection_args = connection_args.copy()
+    # bucket name is an invalid connection argument.
+    connection_args.pop('bucket_name')
+    connection_args.pop('verify_md5')
+    return S3Connection(**connection_args)
+
+
+def upload(source_key, dest_keys, verify_md5=False):
+    """
+    Download from the source s3 key to a named temporary file
+    and upload to the destination s3 key.
+    ``source_key`` The source boto s3 key.
+    ``dest_keys`` A list of the destination boto s3 keys.
+    ``verify_md5`` Verify md5 values when uploading. Source is always
+    considered authoritative.
+    """
+    # Use the same name if no destination key is passed.
+    if not dest_keys or not source_key:
+        raise Exception(
+            'The source_key and dest_keys parameters are required.')
+    with tempfile.NamedTemporaryFile() as data:
+        source_key.get_contents_to_file(data)
+        data.file.flush()
```

Comment:
NamedTemporaryFiles are deleted on close. So the biggest concern, unrecovered resources, seems to be handled appropriately. The NameTemporaryFile scheme allows for `mode='w+b'` which allows them to be opened multiple times yet still be reliably deleted.

If Ceph (and Swift) were more S3 compliant then we could just use the current libraries and everything would be significantly simplified. We could just stream from source to destination.